os/kernel/task/task_event_wait.c

```diff
@@ -0,0 +1,99 @@
+/****************************************************************************
+ *
+ * Copyright 2020 Samsung Electronics All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied. See the License for the specific
+ * language governing permissions and limitations under the License.
+ *
+ ****************************************************************************/
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <assert.h>
+#include <stdbool.h>
+#include <string.h>
+#include <arch/irq.h>
+#include <sys/types.h>
+#include <tinyara/arch.h>
+#include <tinyara/sched.h>
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+static void update_event_queue(struct tcb_s *tcb)
+{
+	if (tcb->pending_event_data != NULL) {
+		tcb->queueing = false;
+		tcb->event_data = tcb->pending_event_data;
+		tcb->pending_event_data = NULL;
+	} else {
+		tcb->event_data = NO_EVENT_DATA;
+	}	
+}
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: task_event_wait
+ *
+ * Description:
+ *   This function is for waiting kernel event.
+ *   If a task calls this function, it will be blocked until receiving event.
+ *
+ * Inputs:
+ *   buf : data buffer for receiving from kernel
+ *   buflen : a length of data buffer
+ *
+ * Return Value:
+ *   On success, OK will be returned.
+ *   On failure, ERROR will be returned.
+ *
+ ****************************************************************************/
+
+int task_event_wait(void *buf, size_t buflen)
+{
+	int ret = ERROR;
+	int saved_state;
+	struct tcb_s *tcb = sched_self();
+	DEBUGASSERT(tcb);
+
+	if (buflen > 0 && buf == NULL) {
+		return ret;
+	}
+
+	saved_state = irqsave();
+	if (tcb->event_data != NO_EVENT_DATA) {
+		/* If there is already received event, return saved event data. */
+		memcpy(buf, tcb->event_data, buflen);
```

Comment:
I think buflen and size of event_data should be checked avoid buffer overflow.

Reply:
Intentionally, I didn't check the size. We will receive only buflen size.